### ISSUE TITLE
MATCH optimization is broken: MATCH does not work with multiple class: statements

### ISSUE THREAD

**Issue by lodrion** [orientechnologies/orientdb#6983]
## OrientDB Version, operating system, or hardware.
- v2.2.12

## Operating System
- [ x] Linux
- [ ] MacOSX
- [ ] Windows
- [ ] Other Unix
- [ ] Other, name? 

## Expected behavior and actual behavior

Adding class: qualifiers to the intermediate clauses in the MATCH query should not cause query to return no results.

## Steps to reproduce the problem

Start with very simple schema with two vertex classes and an edge
```

create class V1 extends V
create class V2 extends V
create class E1 extends E

create vertex V1 set name = 'foo'
create vertex V2 set name = 'bar1'
create vertex V2 set name = 'bar2'

create edge E1 from (SELECT FROM V1) to (SELECT FROM V2)

```

a simple query that roundtrips through E1 works

```

MATCH {
	class: V2,
	as: origin
}.in('E1') {
	as: middle
}.out('E1') {
	as: dest,
	where: (@rid = $matched.origin.@rid)
}
RETURN origin, middle, dest

```

and returns 2 results. 
Adding a class specified to the middle clause 

```

MATCH {
	class: V2,
	as: origin
}.in('E1') {
        class: V1,
	as: middle
}.out('E1') {
	as: dest,
	where: (@rid = $matched.origin.@rid)
}
RETURN origin, middle, dest

```

causes the query to return no results, even through "middle" results from the first query are of class V1.
Removing the where: clause from the third clause causes the query to return (unfiltered) results

```

MATCH {
	class: V2,
	as: origin
}.in('E1') {
        class: V1,
	as: middle
}.out('E1') {
	as: dest
}
RETURN origin, middle, dest

```

And the most interesting thing is that we can make this query to work in a very non-obvious way:
 
```

create vertex V1 set name='foo2'
create vertex V1 set name='foo3'
create vertex V1 set name='foo4'

MATCH {
	class: V2,
	as: origin
}.in('E1') {
        class: V1,
	as: middle
}.out('E1') {
	as: dest,
	where: ($matched.origin.@rid = @rid)
}
RETURN origin, middle, dest

```

And the query has started to work!

Running with the real data of the similar shape, the warning messages that we get when the query fails indicate that this is caused by the query optimizer starting evaluation from the middle clause.

**Comment by obi1kenobi** [comment]
This is related to #6845, both of which are a consequence of not using a topological sort to generate the evaluation order for the query, as discussed in #6931.

**Comment by luigidellaquila** [comment]
Hi @lodrion 

This is already fixed in latest 2.2.14

Closing

Thanks

Luigi
